packages/grid/x-data-grid-generator/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,246 @@
+  } else {
+    firstRowIndex = page * pageSize;
+    lastRowIndex = (page + 1) * pageSize;
+  }
+  const response: FakeServerResponse = {
+    returnedRows: filteredRows.slice(firstRowIndex, lastRowIndex),
+    nextCursor,
+    rowNumber,
+  };
+
+  return new Promise<FakeServerResponse>((resolve) => {
+    setTimeout(() => {
+      resolve(response);
+    }, delay); // simulate network latency
+  });
+};
+
+interface FakeServerResponse {
+  returnedRows: GridRowModel[];
+  nextCursor?: string;
+  rowNumber: number;
+}
+export interface ServerOptions {
+  minDelay: number;
+  maxDelay: number;
+  useCursorPagination?: boolean;
+}
+
+export interface QueryOptions {
+  cursor?: GridRowId;
+  page?: number;
+  pageSize?: number;
+  // TODO: implement the behavior liked to following models
+  filterModel?: GridFilterModel;
+  sortModel?: GridSortModel;
+}
+
+export const serverConfiguration = (
+  dataSetOptions: UseDemoDataOptions,
+  serverOptions: ServerOptions,
+) => {
+  const columns = getColumnsFromOptions(dataSetOptions);
+  const initialState = getInitialState(dataSetOptions, columns);
+
+  const defaultColDef = getGridDefaultColumnTypes();
+  const columnsWithDefaultColDef = columns.map((column) => ({
+    ...defaultColDef[column.type || 'string'],
+    ...column,
+  }));
+
+  const useQuery = (queryOptions: QueryOptions) => {
+    const {
+      data: { rows },
+      loading: dataGenerationIsLoading,
+    } = useDemoData(dataSetOptions);
+
+    const queryOptionsRef = React.useRef(queryOptions);
+    const [rowCount, setRowCount] = React.useState<number | undefined>(undefined);
+    const [nextCursor, setNextCursor] = React.useState<string | undefined>(undefined);
+    const [isLoading, setIsLoading] = React.useState<boolean>(false);
+    const [data, setData] = React.useState<GridRowModel[]>([]);
+
+    React.useEffect(() => {
+      if (dataGenerationIsLoading) {
+        // dataset is not ready
+        return () => {};
+      }
+
+      queryOptionsRef.current = queryOptions;
+      let active = true;
+
+      setIsLoading(true);
+      setRowCount(undefined);
+      setNextCursor(undefined);
+      loadServerRows(rows, queryOptions, serverOptions, columnsWithDefaultColDef).then(
+        ({ returnedRows, nextCursor: responseNextCursor, rowNumber }) => {
+          if (!active) {
+            return;
+          }
+          setData(returnedRows);
+          setIsLoading(false);
+          setRowCount(rowNumber);
+          setNextCursor(responseNextCursor);
+        },
+      );
+
+      return () => {
+        active = false;
+      };
+    }, [dataGenerationIsLoading, queryOptions, rows]);
+
+    // We use queryOptions pointer to be sure that isLoading===true as soon as the options change
+    const effectShouldStart = queryOptionsRef.current !== queryOptions;
+
+    return {
+      isLoading: dataGenerationIsLoading || isLoading || effectShouldStart,
+      data,
+      rowCount,
+      nextCursor,
```

Comment:
With the exception of `isLoading`, a request to an API usually brings back the row count and the next cursor in the same object. If we save this object in the state it means only a single rerender. Instead, here it requires at least 2 more rerenders to update these other values.

![image](https://user-images.githubusercontent.com/42154031/160016069-1e7ad437-21a4-44b0-9864-7ed29f795951.png)

Reply:
If I understand correctly the setState hooks are not batched because they are in a promise

Reply:
Yeah, each call to `setState` will cause a rerender. In https://github.com/mui/mui-x/pull/4186/files/7d68109152a6cc509b3b7305f0c41f9652133117#diff-94d76f5b777c114259dd01edd0339fd1541887e6b5646ad97b14fd1d2fcfec5dR222-R225 we first update the data, then the row count. This may cause that we first get the data without the correct row count. One way to solve this is to save the entire object a single `setState`.